src/material-experimental/mdc-slider/global-change-listener.ts

```diff
@@ -0,0 +1,56 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {DOCUMENT} from '@angular/common';
+import {Inject, Injectable} from '@angular/core';
+import {SpecificEventListener} from '@material/base';
+import {Subject, Subscription} from 'rxjs';
+import {finalize} from 'rxjs/operators';
+
+/**
+ * Handles listening for all change events that occur on the document.
+ *
+ * This service exposes a single method #listen to allow users to subscribe to change events that
+ * occur on the document. Since listening for all change events on the document can be expensive,
+ * we lazily attach a single event listener to the document when the first subscription is made,
+ * and remove the event listener once the last observer unsubscribes.
+ */
+@Injectable({providedIn: 'root'})
+export class GlobalChangeListener {
```

Comment:
I'm not sure that these `change` events are performance-sensitive enough to warrant a document-level listener. I think that this may actually end up being slower, because Zone.js will trigger change detection on the entire document, rather than only the subtree that the `change` event came from.

Reply:
So this is a kind of complicated solution to a somewhat complex problem we need to solve here. I'll do my best to explain here, but if you're still confused at the end, feel free to ping me and lets set up a VC.

#### The problem

MDC uses real native range inputs under the hood for the MDC Slider. Because this is the case, as part of Angular's conventions, we expose these inputs directly to developers through `ng-content` so that the api looks something like this:

```html
<mat-slider>
    <input matSliderThumb>
</mat-slider>
```

As a developer, if I wanted to listen for change or input events on the slider input, I would intuitively do it like this:

```html
<mat-slider>
    <input (change)="onChange($event)" (input)="onInput($event)" matSliderThumb>
</mat-slider>
```

Unfortunately, the MDC implementation does not properly trigger all of the correct events on the native input elements. For change events in particular, using arrow keys will trigger change events, but mouse events will not. However, the MDC implementation _does_ expose two callbacks in the adapter called `#emitChangeEvent` and `#emitInputEvent`, which does correctly tell us when these events are happening.

#### The solution

* Prevent all change and "input" "real" events triggered by interacting with the slider from reaching the user - we'll call these "real" events.

* Dispatch our own "fake" events from the `#emitChangeEvent` and `#emitInputEvent` callbacks which we allow to reach the user.

We want to stop these "real" events from reaching the user. We cannot simply attach an event listener ourselves (for example, in the MatSliderThumb constructor) and call `stopImmediatePropagation` there because the users's event listener has already been attached before we reach the constructor. This means the users's listener will get hit before ours does.

As a workaround, we are attaching the event listener directly to the document, setting `useCapture` to true, and calling `stopImmediatePropagation`.

#### Other info

We do still want these "real" events to reach MDC, which is why we are still calling their `handler` callback in `#registerInputEventHandler`. This is because MDC depends on these for some of its functionality (e.g. updating the ui on arrow key down and calling `emitChangeEvent` & `emitInputEvent`.

We then allow "fake" events to reach users, but make sure not to call the `handler` callback provided by MDC, otherwise it will re-trigger the `#emit<Change or Input>Event` and send us into an infinite loop.